refactor(swarm)!: refactor ListenUpgradeError and DialUpgradeError

jxs opened the pull request:
## Description

Addresses #2894

## Notes

In the protocols, moved the `TimedOut` variants from nested `UpgradeError`'s to `Event` variants themselves. Like `identify` `dcutr` and `relay`. 
**Per commit review is suggested.**

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

jxs committed:
swarm: remove unconstructed Timer variant

from ConnectionHandlerUpgrErr

jxs committed:
swarm: make ListenUpgradeError::error an UpgradeError

instead of ConnectionHandlerUpgrErr, on case of timeout expiration we
just log the error as we don't know which ConnectionHandler should
handle the error.

jxs committed:
swarm: make DialUpgradeError::error an UpgradeError

instead of ConnectionHandlerUpgrErr, and create ConnectionEvent::DialTimeout
for the situations where that upgrading an outbound substream to the given protocol has expired its timeout.

jxs committed:
dcutr: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
identify: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
relay: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
metrics: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
kad: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
rendezvous: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
ping: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates

jxs committed:
gossipsub: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates.

gossipsub/handler: Move dial upgrade error logic from poll to
on_dial_upgrade_error and rename pending_errors to pending_events, so
that handler flow becomes more similar to other protocols.

jxs committed:
request-response: update to libp2p-swarm ListenUpgradeError

and DialUpgradeError updates.

jxs committed:
examples: update file-sharing example to libp2p-swarm

 ListenUpgradeError and DialUpgradeError updates.

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,6 +1,8 @@
 
 - Remove uncontructed variant `Timer` from `ConnectionHandlerUpgrErr`.
+  Make `ListenUpgradeError::error` an `UpgradeError` instead of `ConnectionHandlerUpgrErr`, in case of timeout expiration we
+  just log the error as we don't know which ConnectionHandler should handle the error.
```
These need a reference to this PR.

jxs replied:
yeah, it was done on 11db523b7d094f6ed1a73d32d332d534525c95db

thomaseizinger commented on swarm/src/connection.rs:
```diff
@@ -239,12 +239,17 @@ where
+                        continue;
+                    }
+                    ConnectionHandlerUpgrErr::Timeout => {
+                        log::debug!("Timeout expired during an inbound substream negotiation")
```
```suggestion
                        log::debug!("Failed to negotiate protocol on inbound stream within {seconds}s")
```

That would be more useful IMO.

jxs replied:
`seconds` doesn't seem to exist in this scope, we probably can make `ConnectionHandlerUpgrErr::Timeout` have it . I will implement after clarifying https://github.com/libp2p/rust-libp2p/pull/3307#discussion_r1064264754

thomaseizinger replied:
> `seconds` doesn't seem to exist in this scope, we probably can make `ConnectionHandlerUpgrErr::Timeout` have it . I will implement after clarifying https://github.com/libp2p/rust-libp2p/pull/3307#discussion_r1064264754

Yes, it was only meant to be illustrative on what I'd like to have printed!

thomaseizinger commented on swarm/src/handler.rs:
```diff
@@ -207,6 +207,8 @@ pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IO
     AddressChange(AddressChange<'a>),
     /// Informs the handler that upgrading an outbound substream to the given protocol has failed.
     DialUpgradeError(DialUpgradeError<OOI, OP>),
+    /// Informs the handler that upgrading an outbound substream to the given protocol has expired its timeout.
+    DialTimeout(DialTimeout<OOI>),
```
Would it make sense for `DialUpgradeError` to be an enum with a `Timeout` and `Upgrade` variant?

thomaseizinger replied:
https://github.com/libp2p/rust-libp2p/issues/3268 is related to this. If we agree to remove the `OpenInfo` from `ConnectionHandler` a lot of this code would be simpler because we wouldn't need to track `info` here.

jxs replied:
At first sight, making `DialUpgradeError` an enum with a `Timeout` and `Upgrade` variant sounded similar to what we currently have, where `ConnectionHandlerUpgrErr` then has the variants : https://github.com/libp2p/rust-libp2p/blob/5b8c3d290187b888b2afc8288c8e0b5843ef4964/swarm/src/handler.rs#L340-L343

But what you are suggesting seems to be in fact renaming `ConnectionHandlerUpgradeError` to `DialUpgradeError` and using it in `ConnectionEvent::DialUpgradeError` right? 
https://github.com/libp2p/rust-libp2p/blob/021502ccd20b67a20d99f1af9c2c53ac659e170c/swarm/src/handler.rs#L442-L448

 In fact it probably makes sense (we are in fact reducing one indirection layer) even more sense than having `Timeout` as root variants where I agree with you that they are errors by themselves. Do you also agree @mxinden?

mxinden replied:
Off the top of my head this sounds good to me. I will have to give this more thought with a clear mind to give a good answer.

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -113,10 +113,13 @@ pub use connection::{
+    SubstreamProtocol,
 };
+
+#[deprecated(since = "0.42.0", note = "Shouldn't be required by end users")]
+pub type ConnectionHandlerUpgrErr<TUpgrErr> = handler::ConnectionHandlerUpgrErr<TUpgrErr>;
```
:clap:

thomaseizinger commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -56,14 +57,17 @@ pub enum Event {
     Dial,
     #[error("Failed to establish substream: {0}.")]
-    Handler(ConnectionHandlerUpgrErr<void::Void>),
+    Handler(UpgradeError<void::Void>),
```
Previously, we emitted an error (and thus closed the connection) when we ran into a timeout. Why are we changing this behaviour now?

jxs replied:
Isn't the behaviour the same? In the sense that the connection was closed because of a timeout and we are bubbling it up? But after reading and replying on https://github.com/libp2p/rust-libp2p/pull/3307#discussion_r1064264754 I now see that Timeout makes sense being a variant on the `Error` itself of `DirectConnectionUpgradeFailed` rather than a variant of Event.

thomaseizinger replied:
> Isn't the behaviour the same? In the sense that the connection was closed because of a timeout and we are bubbling it up?

Unless I am mistaken, the timeout refers to a stream upgrade. The connection can/should still be intact at this stage. Something like `PendingUpgrade` should trigger this timeout on the remote but the connection underneath works perfectly fine.

It does raise the question whether we should close an entire connection at all just because one protocol failed to upgrade a stream. Just disabling that protocol (and handler) and letting the collaborative keep-alive eventually close the connection seems like a better idea at first sight.

We may want to do this in a separate PR though, either before or after we merge this.

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -459,8 +462,11 @@ pub enum Event {
     },
+
+    /// Timeout expired while attempting to identify the remote.
+    Timeout,
```
A timeout is an error so it seems odd that we have an `Error` variant and then this. I'd suggest we either log this internally and don't report it or create a dedicated error enum ourselves.

jxs replied:
Since we are already reporting it up I'd go as we do in `dcutr` and expose it as an `Error`, updated to do that! Does it make sense to you to also do the same for `relay`?

thomaseizinger replied:
> Does it make sense to you to also do the same for `relay`?

Yes I think we should apply that thinking to all protocols.

jxs replied:
thanks done, ptal! Also removed what seems to be unrequired `inbound_hop::UpgradeError` and shortcircuit to `inbound_hop::FatalUpgradeError`  see c2eda37b0997e2cf7bc306db2e63796e3c9a069a and see if you agree.

thomaseizinger reviewed (commented):
I love the simplifications coming from this!

I don't quite agree with reporting a kind of error (timeout) next to other errors. I think that should be unified or not reported at all. I am tending towards the latter for a lot of protocols where we can make a decision on what that timeout means.

jxs committed:
dcutr: report Timeout as a variant of Error

instead of as a variant of Event.

jxs committed:
identify: introduce Error enum

and Timeout as part of Error instead of as a variant of Event.

jxs committed:
metrics: fix removal of identify Event::Timeout

jxs committed:
relay: introduce behaviour::OutboundError enum

and Timeout as part of Error instead of as a variant of behaviour::Event.
Introduce `client::InboundError` and Timeout as part of Error instead of as a variant of client::Event.

jxs committed:
relay: remove unneeded inbound_hop::UpgradeError.
